plugin/pkg/admission/security/podsecuritypolicy/admission.go

```diff
@@ -0,0 +1,244 @@
+
+// NewPlugin creates a new PSP admission plugin.
+func NewPlugin(kclient clientset.Interface) *constraint {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return kclient.Extensions().PodSecurityPolicies().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return kclient.Extensions().PodSecurityPolicies().Watch(options)
+			},
+		},
+		&extensions.PodSecurityPolicy{},
+		store,
+		0,
+	)
+
+	return &constraint{
+		Handler: admission.NewHandler(admission.Create),
+		client:  kclient,
+
+		store:     store,
+		reflector: reflector,
+	}
+}
+
+func (a *constraint) Run() {
+	if a.stopChan == nil {
+		a.stopChan = make(chan struct{})
+		a.reflector.RunUntil(a.stopChan)
+	}
+}
+func (a *constraint) Stop() {
+	if a.stopChan != nil {
+		close(a.stopChan)
+		a.stopChan = nil
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available PSPs.
+//
+// 1.  Find available PSPs.
+// 2.  Create the providers, includes setting pre-allocated values if necessary.
+// 3.  Try to generate and validate a PSP with providers.  If we find one then admit the pod
+//     with the validated PSP.  If we don't find any reject the pod and give all errors from the
+//     failed attempts.
+func (c *constraint) Admit(a admission.Attributes) error {
+	if a.GetResource() != api.Resource("pods") {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*api.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
+	glog.V(4).Infof("getting pod security policies for pod %s (generate: %s)", pod.Name, pod.GenerateName)
+	matchedPolicies, err := getMatchingPolicies(c.store)
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+
+	providers, errs := c.createProvidersFromPolicies(matchedPolicies)
+	logProviders(pod, providers, errs)
+
+	if len(providers) == 0 {
+		return admission.NewForbidden(a, fmt.Errorf("no providers available to validate pod request"))
+	}
+
+	// all containers in a single pod must validate under a single provider or we will reject the request
+	validationErrs := field.ErrorList{}
+	for _, provider := range providers {
+		if errs := assignPodSecurityPolicy(provider, pod, field.NewPath(fmt.Sprintf("provider %s: ", provider.GetPSPName()))); len(errs) > 0 {
+			validationErrs = append(validationErrs, errs...)
+			continue
+		}
+
+		// the entire pod validated, annotate and accept the pod
+		glog.V(4).Infof("pod %s (generate: %s) validated against provider %s", pod.Name, pod.GenerateName, provider.GetPSPName())
+		if pod.ObjectMeta.Annotations == nil {
+			pod.ObjectMeta.Annotations = map[string]string{}
+		}
+		pod.ObjectMeta.Annotations[psputil.ValidatedPSPAnnotation] = provider.GetPSPName()
```

Comment:
Is this annotation purely informative?  Or does the security model in OS rely on is particular annotation not being modifiable by normal users?